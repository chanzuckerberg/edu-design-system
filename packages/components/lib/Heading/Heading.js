"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const typography_1 = __importDefault(require("../common/typography"));
const Heading = (0, react_1.forwardRef)((_a, ref) => {
    var { as, children, size } = _a, 
    /**
     * Components that wrap typography sometimes require props such as
     * event handlers, tabIndex, etc. to be passed down into the element.
     * TODO: add better typing or documentation for optional props,
     * e.g. React.HTMLAttributes<HTMLHeadingElement>
     */ rest = __rest(_a, ["as", "children", "size"]);
    return (react_1.default.createElement(typography_1.default, Object.assign({ as: as || size, size: size, ref: ref }, rest), children));
});
Heading.displayName = "Heading"; // Satisfy eslint.
exports.default = Heading;
//# sourceMappingURL=Heading.js.map