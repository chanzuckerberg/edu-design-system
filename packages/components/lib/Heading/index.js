"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = void 0;
var Heading_1 = require("./Heading");
Object.defineProperty(exports, "default", { enumerable: true, get: function () { return __importDefault(Heading_1).default; } });
//# sourceMappingURL=index.js.map