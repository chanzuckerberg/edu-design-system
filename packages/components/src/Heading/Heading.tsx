import React, { ReactNode } from "react";
import {
  TypographyColor,
  TypographySize,
  styleFromColor,
  styleFromSize,
} from "../util/typography";
import tw, { styled } from "twin.macro";

type HeadingElement = "h1" | "h2" | "h3" | "h4" | "h5" | "h6";

type Props = {
  /**
   * This prop can be used to specify which size heading should
   * actually be rendered, in the case that you want to render an element
   * as one heading but style it as if it were another. If both an `as` prop
   * and a `size` prop are passed, the `as` will be used to determine the element
   * and the `size` will be used to determine the styling.
   */
  as?: HeadingElement;
  /**
   * Specifies font weight as either bold or normal (defaults to bold).
   * TODO: Currently we will allow toggling any size, but in the future
   * we will add stricter enforcement -- e.g. enforcing a boldness
   * for each size or checking for mutually exclusive props.
   */
  bold?: boolean;
  /**
   * The text content to present.
   */
  children: ReactNode;
  /**
   * The color of the heading element. If no color provided, defaults to a base color.
   */
  color?: TypographyColor;
  /**
   * The size of the html element. If no `as` prop is provided, then
   * the component uses this value to determine which html tag to render
   * (e.g. 'h1', 'h2', etc.)
   */
  size: TypographySize;
};

interface BaseHeadingProps {
  bold: boolean;
  color?: TypographyColor;
  size: TypographySize;
}

const HeadingComponent = styled.span<BaseHeadingProps>(
  ({ bold, color, size }) => [
    bold ? tw`font-bold` : tw`font-normal`,
    styleFromColor(color),
    styleFromSize(size),
  ]
);

function Heading({
  as,
  bold = true,
  children,
  size,
  /**
   * Components that wrap typography sometimes require props such as
   * event handlers, tabIndex, etc. to be passed down into the element.
   * TODO: add better typing or documentation for optional props,
   * e.g. React.HTMLAttributes<HTMLHeadingElement>
   */
  ...rest
}: Props) {
  return (
    <HeadingComponent as={as || size} bold={bold} size={size} {...rest}>
      {children}
    </HeadingComponent>
  );
}

export default Heading;
