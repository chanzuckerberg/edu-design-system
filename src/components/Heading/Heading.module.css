@import '../../design-tokens/tier-2-usage/typography-usage.css';

/*------------------------------------*\
    #HEADING
\*------------------------------------*/

/**
 * Remove top and bottom margins applied by browser for headings.
 * 1) Needs `:where` pseudoclass to reduce specificity so class overrides 
 *    for margin can happen where used.
 */
:where(.heading) {
  /* 1 */
  margin-top: 0;
  margin-bottom: 0;
}

/* New type ramp sizes */
.heading--size-headline-lg {
  @mixin eds-theme-typography-headline-lg;
}

.heading--size-headline-md {
  @mixin eds-theme-typography-headline-md;
}

.heading--size-headline-sm {
  @mixin eds-theme-typography-headline-sm;
}

.heading--size-title-md {
  @mixin eds-theme-typography-title-md;
}

.heading--size-title-sm {
  @mixin eds-theme-typography-title-sm;
}

/* Deprecated; not in figma */
.heading--size-body-sm {
  @mixin eds-theme-typography-body-text-sm-bold;
}

/* Deprecated; not in figma */
.heading--size-body-xs {
  @mixin eds-theme-typography-body-text-xs-bold;
}

/**
 * Heading sizes, correlate to similar numerically named typography token.
 */
.heading--size-h1 {
  @mixin eds-theme-typography-heading-1;
}

.heading--size-h2 {
  @mixin eds-theme-typography-heading-2;
}

.heading--size-h3 {
  @mixin eds-theme-typography-heading-3;
}

.heading--size-h4 {
  @mixin eds-theme-typography-heading-4;
}

.heading--size-h5 {
  @mixin eds-theme-typography-heading-5;
}

/* Deprecated; not in figma */
.heading--size-h6 {
  @mixin eds-theme-typography-heading-6;
}

/* Deprecated; not in figma */
.heading--size-h7 {
  @mixin eds-theme-typography-heading-7;
}

/**
 * Color variants
 */
.heading--error {
  color: var(--eds-theme-color-text-utility-error);
}
.heading--base {
  color: var(--eds-theme-color-text-neutral-default);
}
.heading--brand {
  color: var(--eds-theme-color-text-brand-primary);
}
.heading--info {
  color: var(--eds-color-info-700);
}
.heading--inherit {
  color: inherit;
}
.heading--neutral {
  color: var(--eds-theme-color-text-neutral-default);
}
.heading--success {
  color: var(--eds-theme-color-text-utility-success);
}
.heading--warning {
  color: var(--eds-theme-color-text-utility-warning);
}
.heading--white {
  color: var(--eds-theme-color-text-neutral-default-inverse);
}
