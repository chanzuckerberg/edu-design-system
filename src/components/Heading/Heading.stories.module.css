/*------------------------------------*\
    # HEADING STORIES
\*------------------------------------*/

/**
 * For the "variant" story to display the different text variant options.
 * 1) Uses grid to neatly align the variant variations.
 */
.variant__table {
  display: grid; /* 1 */
  grid-template-columns: 1fr 1fr 1fr;
  gap: var(--eds-size-half);
}

/**
 * Color variants are segmented into left, middle, and right columns.
 * Backgrounds were chosen to represent typical body backgrounds.
 */
.variant--white {
  grid-column: 1;
  background-color: var(--eds-theme-color-background-neutral-default);
}

.variant--light {
  grid-column: 2;
  background-color: var(--eds-theme-color-background-neutral-default);
}

.variant--dark {
  grid-column: 3;
  background-color: var(--eds-theme-color-background-neutral-default-inverted);
}
