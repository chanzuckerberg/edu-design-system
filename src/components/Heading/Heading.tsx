import { clsx } from 'clsx';
import React, { forwardRef } from 'react';
import styles from './Heading.module.css';

export const VARIANTS = [
  'inherit',
  'neutral-subtle',
  'neutral-medium',
  'neutral-strong',
  'brand',
  'success',
  'warning',
  'error',
  'white',
  /**
   * @deprecated Info variant is deprecated.
   */ 'info',
] as const;
export type Variant = (typeof VARIANTS)[number];
export type HeadingElement = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6';
// For now, "h1"-"h6" sizes point to the old type ramp, while
// "headline-*" and "title-*" sizes point to the new type ramp.
// These will be brought in sync with the next major release.
const TOKEN_TO_SIZE = {
  'headline-lg': 'h1',
  'headline-md': 'h2',
  'headline-sm': 'h3',
  'title-md': 'h4',
  'title-sm': 'h5',
  'body-sm': 'h6',
  'body-xs': 'h7',
  'title-xs': 'h5',
};
export type HeadingSize = HeadingElement | 'h7' | keyof typeof TOKEN_TO_SIZE;

type Props = {
  /**
   * This prop can be used to specify which size heading should
   * actually be rendered, in the case that you want to render an element
   * as one heading but style it as if it were another. If both an `as` prop
   * and a `size` prop are passed, the `as` will be used to determine the element
   * and the `size` will be used to determine the styling.
   */
  as?: HeadingElement;
  children: React.ReactNode;
  className?: string;
  size: HeadingSize;
  tabIndex?: number;
  variant?: Variant;
} & React.HTMLAttributes<HTMLHeadingElement>;

const getComputedAs = (size: HeadingSize) => {
  if (size === 'h7') {
    return 'h6';
  }

  return size in TOKEN_TO_SIZE ? TOKEN_TO_SIZE[size] : size;
};

/**
 * `import {Heading} from "@chanzuckerberg/eds";`
 *
 * A component for styling heading text (`<h1>`-`<h6>`).
 *
 * Be careful to pass the correct heading element via the `as` prop to avoid skipping heading levels because [heading levels increasing by only one level at a time is important for screen reader users.](https://www.w3.org/WAI/tutorials/page-structure/headings/)
 */
export const Heading = forwardRef(
  (
    {
      as,
      children,
      className,
      variant,
      size,
      /**
       * Components that wrap typography sometimes require props such as
       * event handlers, tabIndex, etc. and/or other native heading element
       * attributes to be passed down into the element.
       */ ...other
    }: Props,
    ref: React.ForwardedRef<HTMLHeadingElement>,
  ) => {
    if (process.env.NODE_ENV !== 'production') {
      if (variant === 'info') {
        console.warn(
          'Info variant is deprecated and will be removed in an upcoming release. Please use the consider another variant instead.',
        );
      }
      if (size === 'h6' || size === 'h7') {
        console.warn(
          `The ${size} size is deprecated and will be removed in an upcoming release.\n`,
          'Please bump this heading up to a larger size if possible.',
        );
      }
    }

    const TagName = as || getComputedAs(size);
    const componentClassName = clsx(
      styles['heading'],
      styles[`heading--size-${size}`],
      variant && styles[`heading--${variant}`],
      className,
    );
    return (
      <TagName className={componentClassName} ref={ref} {...other}>
        {children}
      </TagName>
    );
  },
);

Heading.displayName = 'Heading'; // Satisfy eslint.
