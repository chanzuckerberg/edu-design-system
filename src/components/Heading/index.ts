export { Heading as default } from './Heading';
export type { HeadingElement, HeadingSize } from './Heading';
