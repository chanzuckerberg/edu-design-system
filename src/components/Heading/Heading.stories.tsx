import { StoryObj } from '@storybook/react';
import React from 'react';
import Heading, { VARIANTS } from './Heading';
import styles from './Heading.stories.module.css';

export default {
  title: 'Atoms/Text/Heading',
  component: Heading,
};

type Args = React.ComponentProps<typeof Heading>;

export const Heading1: StoryObj<Args> = {
  args: {
    size: 'h1',
    children: 'Heading 1',
  },
};

export const Heading2: StoryObj<Args> = {
  args: {
    size: 'h2',
    children: 'Heading 2',
  },
};

export const Heading3: StoryObj<Args> = {
  args: {
    size: 'h3',
    children: 'Heading 3',
  },
};

export const Heading4: StoryObj<Args> = {
  args: {
    size: 'h4',
    children: 'Heading 4',
  },
};

export const Heading5: StoryObj<Args> = {
  args: {
    size: 'h5',
    children: 'Heading 5',
  },
};

export const Heading6: StoryObj<Args> = {
  args: {
    size: 'h6',
    children: 'Heading 6',
  },
};

export const Variants: StoryObj<Args> = {
  ...Heading1,
  render: (args) => {
    const headings = [];
    VARIANTS.forEach((variant) => {
      if (variant !== 'info')
        headings.push(
          <Heading {...args} variant={variant}>
            {variant}
          </Heading>,
        );
    });
    return (
      <div className={styles['variant__table']}>
        <div className={styles['variant--white']}>{headings}</div>
        <div className={styles['variant--light']}>{headings}</div>
        <div className={styles['variant--dark']}>{headings}</div>
      </div>
    );
  },
};

export const Sizes: StoryObj<Args> = {
  render: () => {
    const sizes = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'] as const;
    const combinations = sizes.map((as: typeof sizes[number]) => {
      const headings = sizes.map((size: typeof sizes[number]) => (
        <Heading as={as} size={size} key={`as${as}size${size}`}>
          as: {as} size: {size}
        </Heading>
      ));
      return <div key={as}>{headings}</div>;
    });
    return <div className={styles['variant__table']}>{combinations}</div>;
  },
};
