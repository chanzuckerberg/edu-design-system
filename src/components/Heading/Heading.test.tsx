import { generateSnapshots } from '@chanzuckerberg/story-utils';
import { render, screen } from '@testing-library/react';

import userEvent from '@testing-library/user-event';
import React from 'react';
import { Heading } from './Heading';
import * as HeadingStoryFile from './Heading.stories';

describe('<Heading />', () => {
  generateSnapshots(HeadingStoryFile);

  it('should add the passthrough className', () => {
    render(
      <Heading size="h1" className="passthrough">
        Some Heading
      </Heading>,
    );
    expect(screen.getByText('Some Heading').classList).toContain('passthrough');
  });
  it('should handle refs', async () => {
    const HelperComponent = () => {
      const refContainer = React.useRef(null);
      const onButtonClick = () => {
        expect(refContainer.current).toBe(
          screen.getByText('Ref container parent test'),
        );
      };
      return (
        <>
          <Heading size="h1" ref={refContainer}>
            Ref container parent test
          </Heading>
          <button onClick={onButtonClick}>Test ref</button>
        </>
      );
    };
    render(<HelperComponent />);
    expect(screen.getByText('Ref container parent test')).toBeTruthy();
    userEvent.click(screen.getByRole('button'));
  });
});
