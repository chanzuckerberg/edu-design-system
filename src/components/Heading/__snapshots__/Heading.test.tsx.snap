// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> Default story renders snapshot 1`] = `
<h1
  class="heading heading--headline-lg"
>
  Default Heading
</h1>
`;

exports[`<Heading /> Heading1 story renders snapshot 1`] = `
<h1
  class="heading heading--headline-lg"
>
  Heading 1
</h1>
`;

exports[`<Heading /> Heading2 story renders snapshot 1`] = `
<h2
  class="heading heading--headline-md"
>
  Heading 2
</h2>
`;

exports[`<Heading /> Heading3 story renders snapshot 1`] = `
<h3
  class="heading heading--headline-sm"
>
  Heading 3
</h3>
`;

exports[`<Heading /> Heading4 story renders snapshot 1`] = `
<h4
  class="heading heading--title-md"
>
  Heading 4
</h4>
`;

exports[`<Heading /> Heading5 story renders snapshot 1`] = `
<h5
  class="heading heading--title-sm"
>
  Heading 5
</h5>
`;

exports[`<Heading /> Heading6 story renders snapshot 1`] = `
<h6
  class="heading heading--title-xs"
>
  Heading 6
</h6>
`;

exports[`<Heading /> UsingColorTokens story renders snapshot 1`] = `
<div>
  <h1
    class="heading heading--title-md text-utility-warning"
  >
    using 
    <code>
      text-utility-warning
    </code>
     utility class
  </h1>
  <h2
    class="heading heading--title-md text-utility-favorable"
  >
    using 
    <code>
      text-utility-favorable
    </code>
     utility class and preset override
  </h2>
  <h1
    class="heading heading--title-md text-utility-critical"
  >
    using 
    <code>
      text-utility-critical
    </code>
     utility class
  </h1>
  <h1
    class="heading heading--title-md text-[var(--eds-theme-color-text-utility-favorable)]"
  >
    using color with token in utility class and preset override
  </h1>
</div>
`;
